src/celeritas/global/detail/ActionSequence.cc

```diff
@@ -103,40 +103,51 @@ template<MemSpace M>
 void ActionSequence<Params>::execute(Params const& params, State<M>& state)
 {
     [[maybe_unused]] Stream::StreamT stream = nullptr;
-    if (M == MemSpace::device && options_.sync)
+    if (M == MemSpace::device && options_.action_times)
     {
         stream = celeritas::device().stream(state.stream_id()).get();
     }
 
-    if ((M == MemSpace::host || options_.sync) && !state.warming_up())
+    // Running a single track slot on host:
+    // Skip inapplicable post-step action
+    auto const skip_post_action = [&](auto const& action) {
+        if constexpr (M != MemSpace::host)
+        {
+            return false;
+        }
+        return state.size() == 1 && action->order() == ActionOrder::post
+               && action->action_id()
+                      != state.ref().sim.post_step_action[TrackSlotId{0}];
+    };
+
+    if (options_.action_times && !state.warming_up())
     {
         // Execute all actions and record the time elapsed
         for (auto i : range(actions_.size()))
         {
-            // if we're running on track slot on host, we can skip unnecessary
-            // post-step actions
-            if (M == MemSpace::host && state.size() == 1
-                && actions_[i]->order() == ActionOrder::post
-                && actions_[i]->action_id()
-                       != state.ref().sim.post_step_action[TrackSlotId{0}])
-                continue;
-            ScopedProfiling profile_this{actions_[i]->label()};
-            Stopwatch get_time;
-            actions_[i]->execute(params, state);
-            if (M == MemSpace::device)
+            if (auto const& action = actions_[i]; !skip_post_action(action))
```

Comment:
Since these are shared pointers I think it would improve readability (and maybe micro-performance by reducing indirection?) to dereference the shared pointer and have `action` be the actual `Action` object rather than a shared pointer:
```suggestion
            if (auto const& action = *actions_[i]; !skip_post_action(action))
```